src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs

```diff
@@ -517,6 +522,13 @@ private ImmutableArray<ISymbol> GetUnnamedSymbols(ExpressionSyntax originalExpre
             return symbols.ToImmutable();
         }
 
+        private ITypeSymbol? GetContainerForUnnamedSymbols(SemanticModel semanticModel, ExpressionSyntax originalExpression)
```

Comment:
I still find my approach in #52189 more suitable. Why do you ignore the hard acquired `container` just to do the same work here again (and facing the same problems)? Passing the container would lead to symmetry between named (`GetMemberSymbols`) and unnamed symbol (`GetUnnamedSymbols`) acquisition:

https://github.com/dotnet/roslyn/blob/cb6cf9c792c844d1b617a58e475023d91402ba89/src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs#L483-L494

(Note: Doing `var container = containerSymbol.GetSymbolType();` works pretty well in #52189)

Do I miss something? Is there something wrong about the `container` that prevents it from being re-used here?

Just my two cents.

Reply:
Because the acquired container symbol is what is wanted for the basic dotted members, not the unnamed members. For example, for `?.` we want the container to be the unwrapped type. However, for operator/conversions, we want it to be the nullable type.

Reply:
I see. But the case you describe, is actually working even if we pass the container with the nulllable removed, because you make sure to add it for null conditional access expressions in `GetUnnamedSymbols`:
https://github.com/dotnet/roslyn/blob/f52afd963b21a472d78d5fe17fec8e65fbd10f74/src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs#L507-L511

That's why my approach was working in the first place.

I took the chance to also look into the other places, where the recommender takes special care of the container type. I added some test in my PR (2d30d1a7) to verify that there are no bugs buried somewhere in there (spoiler: it's not). This is what I looked into:

### `GetSymbolsOffOfName`

This is the case described in the issue. Fixed now by speculative binding.

### `GetSymbolsOffOfExpression`

`container` is `semanticModel.GetTypeInfo(expression, _cancellationToken).Type`, so nothing special happens here in the first step. But a special handling is done in the second step for the `Color Color` case (property name and type name are equivalent), but this does not affect the unnamed symbols. I added a test in my PR anyway.

### `GetSymbolsOffOfConditionalReceiver`

That's the case described by you. The nullable is removed from the type (`var container = _context.SemanticModel.GetTypeInfo(expression, _cancellationToken).Type.RemoveNullableIfPresent();`). But as mentioned earlier, you put nullable back anyway before you collect the unnamed symbols.

### `GetSymbolsOffOfDereferencedExpression`

The `p->$$` case. The provider only kicks in after `.`,  so it is not triggered. This is the right behavior and I added a test for this.

From my point of view, both PRs solve the problem, so it's up to you which way you prefer. I would recommend including the tests of commit 2d30d1a7 into the final PR that get's merged into main.

Reply:
i prefer my way :)